src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -139,226 +136,27 @@ public void Add(Operation operation, object objectToApplyTo)
-                            array.Add(conversionResult.ConvertedInstance);
-                            treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(
-                                treeAnalysisResult.PropertyPathInParent, array);
-                        }
-                        else
-                        {
-                            // specified index must not be greater than 
-                            // the amount of items in the array
-                            if (positionAsInteger > array.Count)
-                            {
-                                LogError(new JsonPatchError(
-                                    objectToApplyTo,
-                                    operationToReport,
-                                    Resources.FormatInvalidIndexForArrayProperty(
-                                        operationToReport.op,
-                                        path)));
-                                return;
-                            }
-
-                            array.Insert(positionAsInteger, conversionResult.ConvertedInstance);
-                            treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(
-                                treeAnalysisResult.PropertyPathInParent, array);
-                        }
-                    }
-                    else
-                    {
-                        // get the actual type
-                        var typeOfPathProperty = treeAnalysisResult.Container
-                            .GetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent).GetType();
-
-                        // can the value be converted to the actual type?
-                        var conversionResult = ConvertToActualType(typeOfPathProperty, value);
-                        if (conversionResult.CanBeConverted)
-                        {
-                            treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(
-                                    treeAnalysisResult.PropertyPathInParent,
-                                    conversionResult.ConvertedInstance);
-                        }
-                        else
-                        {
-                            LogError(new JsonPatchError(
-                               objectToApplyTo,
-                               operationToReport,
-                               Resources.FormatInvalidValueForProperty(conversionResult.ConvertedInstance, path)));
-                            return;
-                        }
-                    }
-                }
-                else
-                {
-                    // New property - add it.  
-                    treeAnalysisResult.Container.Add(treeAnalysisResult.PropertyPathInParent, value);
-                }
+                var patchObject = ObjectVisitor.Visit(context);
+                patchObject.Add(value);
             }
-            else
+            catch (JsonPatchException exception)
```

Comment:
This could also use `catch (...) when (...)` which would be an improvement in clarity as well as provide essential PranavPoints.